src/Spago/Build.hs

```diff
@@ -78,21 +80,36 @@ build BuildOptions{..} maybePostBuild = do
   case noInstall of
     DoInstall -> Fetch.fetchPackages cacheConfig deps packagesMinPursVersion
     NoInstall -> pure ()
-  let allGlobs = Packages.getGlobs deps depsOnly configSourcePaths <> sourcePaths
-      buildAction = do
-        Purs.compile allGlobs pursArgs
+
+  let allPsGlobs = Packages.getGlobs   deps depsOnly configSourcePaths <> sourcePaths
+      allJsGlobs = Packages.getJsGlobs deps depsOnly configSourcePaths <> sourcePaths
+      buildAction globs = do
+        Purs.compile globs pursArgs
         case maybePostBuild of
           Just action -> action
           Nothing     -> pure ()
-  absoluteGlobs <- traverse makeAbsolute $ Text.unpack . Purs.unSourcePath <$> allGlobs
-  absoluteJSGlobs <- traverse makeAbsolute $ Text.unpack . Purs.unSourcePath
-    <$> (Packages.getJsGlobs deps depsOnly configSourcePaths <> sourcePaths)
+
+  absoluteJSGlobs <- makeGlobsAbsolute allJsGlobs
 
   case shouldWatch of
-    BuildOnce -> buildAction
-    Watch     -> Watch.watch
-      (Set.fromAscList $ fmap Glob.compile $ absoluteGlobs <> absoluteJSGlobs)
-      shouldClear buildAction
+    BuildOnce -> buildAction allPsGlobs
+    Watch -> do
+      matches <- filterMatchingGlobsAndWarn allPsGlobs
```

Comment:
Why do we match only on PS globs but not the JS ones?

Reply:
Oh I see, I guess we do it only on the PS ones because we care about the directory being there, which we would detect anyways. I'd like to have a comment on top of this line explaining what I just figured out though 😄

Reply:
@f-f I only just pulled in the JS glob changes from `master` and was pondering this behaviour too 😅 I think it might be better to filter the JS globs too in case they end up differing for some reason?

Reply:
It might make sense, though you'd end up with "duplicated" errors, right?

Alternatively we could group all the "non matching" globs and spit out a giant warning with all the non-matching paths after all the filtering

Reply:
@f-f So, while playing with warning for JS files I noticed some interesting behaviour.

`spago.dhall`:

```
sources = [ "src/**/*.purs", "test/**/*.purs", "fake/**/*.js" ]
```

Partial output from `spago build -w`:

```
WARNING: No matching files found when trying to watch pattern: test/**/*.purs
WARNING: No matching files found when trying to watch pattern: fake/**/*.js
WARNING: No matching files found when trying to watch pattern: ./../package-b/src/**/*.js
WARNING: No matching files found when trying to watch pattern: .spago/psci-support/v4.0.0/src/**/*.js
WARNING: No matching files found when trying to watch pattern: src/**/*.js
WARNING: No matching files found when trying to watch pattern: test/**/*.js
WARNING: No matching files found when trying to watch pattern: fake/**/*.js
```

A few observations:

1. There are separate warnings for `test/**/*.purs` and `test/**/*.js`; as you say, we could aggregate these
2. There's a warning for `src/**/*.js` even though there are `.purs` files in that directory, which feels like bad UX
3. `fake/**/*.js` shows up twice because we currently assume all the globs in `spago.dhall` are for `.purs` files
4. You get warnings for `package-b` (a local dependency from the monorepo) and `.spago/psci-support`. We probably shouldn't be watching anything in `.spago`, but either way I don't think it would be expected behaviour for these warnings to show up here

1 and 2 can be addressed by checking that the directory exists rather than the specific types of files inside, but 3 and 4 might need some further consideration.

Reply:
@elliotdavies excellent analysis! 👏 

Interesting corner cases in here. For (1) and (2) we should indeed just convert globs to their parent folder and check that they exist. I think we can reuse [`globToParent`](https://github.com/spacchetti/spago/blob/46f0f7c3f0d3c630b5d03e9f6ef073caa847402a/src/Spago/Watch.hs#L178) for this (we do use it already in the watcher, which means that we might be able to consolidate the check from this PR inside the `setWatched` in there

To solve (3) I think we should also get the parent directory of the glob, and generate both the `.js` and `.purs` globs from there (we can fix this here or just log it in an issue and refactor the glob handling later)

Then (4) is a thing because of https://github.com/spacchetti/spago/issues/172, but you're right that we could avoid watching stuff in `.spago` and only watch local dependencies

Reply:
@f-f OK great, I've raised these as separate issues because they're not really related to this PR's intention:

- #431 for (3) and to think about globs in `spago.dhall` more generally
- #430 to avoid watching dependencies in `.spago` for efficiency and to mitigate (4)

For (1) and (2) I'd been thinking along the same lines so I'll go ahead and do that 😄